File: esports-backend/src/main/java/com/esports/service/DashboardService.java
package com.esports.service;

import com.esports.dto.DashboardResponse;
import com.esports.entity.Project;
import com.esports.entity.User;
import com.esports.repository.ProjectRepository;
import com.esports.repository.ReviewRepository;
import com.esports.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@Service
public class DashboardService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private ProjectRepository projectRepository;

    @Autowired
    private ReviewRepository reviewRepository;

    public DashboardResponse getCreatorDashboard(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new NoSuchElementException("User not found"));

        List<Project> projects = projectRepository.findByCreatorOrderByCreatedAtDesc(user);

        long activeProjects = projects.stream()
                .filter(p -> p.getStatus() != Project.ProjectStatus.COMPLETED).count();
        long completedProjects = projects.stream()
                .filter(p -> p.getStatus() == Project.ProjectStatus.COMPLETED).count();
        double totalSpent = projects.stream()
                .filter(p -> p.getPaymentStatus() == Project.PaymentStatus.COMPLETED)
                .mapToDouble(Project::getPrice).sum();

        DashboardResponse.DashboardStats stats = new DashboardResponse.DashboardStats(
                projects.size(), activeProjects, completedProjects, totalSpent, 0.0
        );

        List<DashboardResponse.ProjectSummary> projectList = projects.stream()
                .map(project -> new DashboardResponse.ProjectSummary(
                        project.getId(),
                        project.getTitle(),
                        formatStatus(project.getStatus().toString()),
                        project.getFreelancer().getName(),
                        project.getPrice()
                ))
                .collect(Collectors.toList());

        return new DashboardResponse(stats, projectList);
    }

    public DashboardResponse getFreelancerDashboard(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new NoSuchElementException("User not found"));

        List<Project> projects = projectRepository.findByFreelancerOrderByCreatedAtDesc(user);

        long activeProjects = projects.stream()
                .filter(p -> p.getStatus() != Project.ProjectStatus.COMPLETED).count();
        long completedProjects = projects.stream()
                .filter(p -> p.getStatus() == Project.ProjectStatus.COMPLETED).count();
        double totalEarnings = projects.stream()
                .filter(p -> p.getStatus() == Project.ProjectStatus.COMPLETED &&
                        p.getPaymentStatus() == Project.PaymentStatus.COMPLETED)
                .mapToDouble(Project::getPrice).sum();

        // Calculate average rating from reviews
        Double avgRating = reviewRepository.findAverageRatingByFreelancer(user);
        double rating = avgRating != null ? Math.round(avgRating * 10.0) / 10.0 : 0.0;

        DashboardResponse.DashboardStats stats = new DashboardResponse.DashboardStats(
                projects.size(), activeProjects, completedProjects, totalEarnings, rating
        );

        List<DashboardResponse.ProjectSummary> projectList = projects.stream()
                .map(project -> new DashboardResponse.ProjectSummary(
                        project.getId(),
                        project.getTitle(),
                        formatStatus(project.getStatus().toString()),
                        project.getCreator().getName(),
                        project.getPrice()
                ))
                .collect(Collectors.toList());

        return new DashboardResponse(stats, projectList);
    }

    private String formatStatus(String status) {
        return switch (status) {
            case "IN_PROGRESS" -> "In Progress";
            case "IN_REVIEW" -> "In Review";
            case "COMPLETED" -> "Completed";
            default -> status.replace("_", " ");
        };
    }
}

File: esports-backend-dotnet/Upkart1/Services/DashboardService.cs
using Microsoft.EntityFrameworkCore;
using Upkart1.Data;
using Upkart1.DTO;
using Upkart1.Entities;
using Upkart1.Entities.Enums;

namespace Upkart1.Services
{
    public class DashboardService : IDashboardService
    {
        private readonly ApplicationDbContext _context;

        public DashboardService(ApplicationDbContext context)
        {
            _context = context;
        }

        public DashboardResponse GetCreatorDashboard(string email)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");
            
            var projects = _context.Projects
                .Include(p => p.Freelancer)
                .Where(p => p.CreatorId == user.Id)
                .OrderByDescending(p => p.CreatedAt)
                .ToList();
            
            var activeProjects = projects.Count(p => p.Status != ProjectStatus.COMPLETED);
            var completedProjects = projects.Count(p => p.Status == ProjectStatus.COMPLETED);
            var totalSpent = projects
                .Where(p => p.PaymentStatus == PaymentStatus.COMPLETED)
                .Sum(p => p.Price);
            
            var stats = new DashboardResponse.DashboardStats(
                projects.Count, activeProjects, completedProjects, totalSpent, 0.0
            );
            
            var projectList = projects.Select(project => new DashboardResponse.ProjectSummary(
                project.Id,
                project.Title,
                FormatStatus(project.Status.ToString()),
                project.Freelancer?.Name ?? "",
                project.Price
            )).ToList();
            
            return new DashboardResponse(stats, projectList);
        }

        public DashboardResponse GetFreelancerDashboard(string email)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");
            
            var projects = _context.Projects
                .Include(p => p.Creator)
                .Where(p => p.FreelancerId == user.Id)
                .OrderByDescending(p => p.CreatedAt)
                .ToList();
            
            var activeProjects = projects.Count(p => p.Status != ProjectStatus.COMPLETED);
            var completedProjects = projects.Count(p => p.Status == ProjectStatus.COMPLETED);
            var totalEarnings = projects
                .Where(p => p.Status == ProjectStatus.COMPLETED && 
                           p.PaymentStatus == PaymentStatus.COMPLETED)
                .Sum(p => p.Price);
            
            // Calculate average rating from reviews
            var avgRating = _context.Reviews
                .Include(r => r.Project)
                .Where(r => r.Project.FreelancerId == user.Id)
                .Average(r => (double?)r.Rating) ?? 0.0;
            var rating = Math.Round(avgRating * 10.0) / 10.0;
            
            var stats = new DashboardResponse.DashboardStats(
                projects.Count, activeProjects, completedProjects, totalEarnings, rating
            );
            
            var projectList = projects.Select(project => new DashboardResponse.ProjectSummary(
                project.Id,
                project.Title,
                FormatStatus(project.Status.ToString()),
                project.Creator?.Name ?? "",
                project.Price
            )).ToList();
            
            return new DashboardResponse(stats, projectList);
        }
        
        private string FormatStatus(string status)
        {
            return status switch
            {
                "IN_PROGRESS" => "In Progress",
                "IN_REVIEW" => "In Review",
                "COMPLETED" => "Completed",
                "PENDING" => "Pending",
                "CANCELLED" => "Cancelled",
                _ => status.Replace("_", " ")
            };
        }
    }
}
